components/message_identifier_router.jsx

```diff
@@ -0,0 +1,218 @@
+    const username = identifier.slice(1, identifier.length).toLowerCase();
+
+    let user = UserStore.getProfileByUsername(username);
+    if (!user) {
+        const {data, error} = await getUserByUsername(username)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    openDirectChannelToUser(
+        user.id,
+        (channel) => {
+            doChannelChange(channel);
+        },
+        () => handleError(match, history)
+    );
+}
+
+async function goToDirectChannelByUserId(match, history) {
+    const {team, identifier} = match.params;
+    const userId = identifier.toLowerCase();
+
+    let user = UserStore.getProfile(userId);
+    if (!user) {
+        const {data, error} = await getUser(userId)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    history.replace(`/${team}/messages/@${user.username}`);
+}
+
+async function goToDirectChannelByUserIds(match, history) {
+    const {team, identifier} = match.params;
+    const userId = Utils.getUserIdFromChannelId(identifier.toLowerCase());
+
+    let user = UserStore.getProfile(userId);
+    if (!user) {
+        const {data, error} = await getUser(userId)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    history.replace(`/${team}/messages/@${user.username}`);
+}
+
+async function goToDirectChannelByEmail(match, history) {
+    const {team, identifier} = match.params;
+    const email = identifier.toLowerCase();
+
+    let user = UserStore.getProfileByEmail(email);
+    if (!user) {
+        const {data, error} = await getUserByEmail(email)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    history.replace(`/${team}/messages/@${user.username}`);
+}
+
+async function goToGroupChannelByGroupId(match, history) {
+    const {identifier} = match.params;
+    const groupId = identifier.toLowerCase();
+
+    history.replace(match.url.replace('/channels/', '/messages/'));
+
+    let channel = ChannelStore.getByName(groupId);
+    if (!channel) {
+        const {data, error} = await joinChannel(UserStore.getCurrentId(), TeamStore.getCurrentId(), null, groupId)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        channel = data.channel;
+    }
+
+    doChannelChange(channel);
+}
+
+function doChannelChange(channel) {
+    GlobalActions.emitChannelClickEvent(channel);
+}
+
+function handleError(match, history) {
+    const {team} = match.params;
+    history.push(team ? `/${team}/channels/${Constants.DEFAULT_CHANNEL}` : '/');
+}
+
+export default class NeedsTeam extends React.PureComponent {
```

Comment:
NeedsTeam? You mean MessageIdentifierRouter?

Reply:
Right? I've wondered about why the previous is

I will change

Reply:
@crspeller I have changed that to ChannelIdentiferRouter on latest commit